### PR TITLE
[algo] feat: add DPPO with binary TV or binary KL implementation

### PR DESCRIPTION
**Pull request by QPHutu**
### What does this PR do?

This PR adds the official implementation of [DPPO](https://arxiv.org/pdf/2602.04879). It replaces the heuristic ratio clipping with a more principled constraint based on a direct estimation of policy divergence (e.g. Total Variation or KL).

Reference Paper: [Rethinking the Trust Region in LLM Reinforcement Learning](https://arxiv.org/pdf/2602.04879)
Original Repo: [Stable-RL](https://github.com/sail-sg/Stable-RL)

The results of running Qwen3-30B-A3B-Base on DAPO dataset:

image

image

image

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [x] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [x] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/trainer/ppo/core_algos.py [hunk 1]
```diff
@@ -1259,6 +1259,172 @@ def compute_policy_loss_vanilla(
     return pg_loss, pg_metrics
 
 
+@register_policy_loss("dppo_tv")
```
[the remark this answers is not in the document]
Comment (tongyx361) [reply verl/trainer/ppo/core_algos.py:1262]: @QPHutu Do you think this suggestion makes sense? If so, let's also use "dppo_binary_kl" for the binary KL function.

[the remark this answers is not in the document]
Comment (QPHutu) [reply verl/trainer/ppo/core_algos.py:1262]: Thanks for your time and comments!

In the paper, we implement too approximation methods: binary and top_k. 

Top_k variant involves non-trivial code changes and potential large latency overhead, especially when TP>1. Our empirical results show that the binary variants perform equally well with top_k. 

Based on these observation, we don't have any plan to support top_k. The binary one is already good enough, and the code change is super simple. Given this, I think no need to use "dppo_binary_tv" and "dppo_tv" should be good.

[the remark this answers is not in the document]
Comment (tongyx361) [reply verl/trainer/ppo/core_algos.py:1262]: Thanks for your explanation. I also think it makes sense to take the binary approximation as the default.

**Review by tongyx361 — DISMISSED** [review]

**Comment by tongyx361** [comment]
@QPHutu Please fix the documentation CI by adding `algo/dppo.md` to `index.rst` and Last updated to `algo/dppo.md`.

**Comment by QPHutu** [comment]
> @QPHutu Please fix the documentation CI by adding `algo/dppo.md` to `index.rst` and Last updated to `algo/dppo.md`.

Hi @tongyx361 , the doc was fixed. Additionally, I changed the `loss_agg_mode` from `seq-mean-token-sum-norm` to `seq-mean-token-sum` to avoid error because in latest verl `seq-mean-token-sum-norm` must be together with a `loss_scale_factor`.

**Comment by tongyx361** [comment]
> > @QPHutu Please fix the documentation CI by adding `algo/dppo.md` to `index.rst` and Last updated to `algo/dppo.md`.
> 
> Hi @tongyx361 , the doc was fixed. Additionally, I changed the `loss_agg_mode` from `seq-mean-token-sum-norm` to `seq-mean-token-sum` to avoid error because in latest verl `seq-mean-token-sum-norm` must be together with a `loss_scale_factor`.

Since the last commit has passed most CI tests, we can skip running again.

**Review by tongyx361 — APPROVED** [review]

**Comment by startju** [comment]
Hi @QPHutu @tongyx361, a follow-up question about a design choice in `compute_policy_loss_dppo_tv` / `compute_policy_loss_dppo_kl`:

```python
pg_losses = -advantages * truncated_ratio * log_prob * valid_mask
```

This is mathematically gradient-equivalent to the straight-through form used in the standard PPO surrogate:

```python
pg_losses = -advantages * truncated_ratio * (1 + log_prob - log_prob.detach()) * valid_mask
```

Both produce identical `∇θ = -A · truncated_ratio · ∇log π`, so training dynamics are the same. The difference is in the **forward loss value**:

- Current form: `pg_loss ≈ -A · ratio · log_prob`. Magnitude scales with `|log_prob|` (typically 1~10) and is entangled with policy entropy.
- Straight-through form: `pg_loss ≈ -A · ratio`. Magnitude hovers near 0, matching the standard PPO surrogate convention.

The concern is **observability when this loss gets reused in a setup with a value head**:

```
total = pg_loss + vf_coef · vf_loss - ent_coef · entropy
```

With the current form:
1. `pg_loss`'s magnitude dominates `total_loss`, so tuning `vf_coef` by visually matching magnitudes is misleading.
2. `total_loss` curves are not comparable to standard PPO implementations.
3. Divergence diagnosis from `total_loss` alone becomes harder (can't tell pg vs vf at a glance).

DPPO's typical use case (GRPO-style, no critic) hides this, but as the loss gets adapted to other algorithms it may surface.

Was the current form a deliberate choice (REINFORCE-style readability, closer to the paper's notation), or just the natural first writing? If the latter, would you be open to switching to the straight-through form for monitoring consistency with the rest of the PPO family in verl?

(Disclosure: this question was drafted with AI assistance.)
